lib/spack/spack/solver/concretize.lp

```diff
@@ -1570,6 +1570,35 @@ compiler_used_as_a_library(node(X, Child), Hash) :-
     attr("virtual_on_edge", PackageNode, CompilerNode2, "cxx"),
     CompilerNode1 != CompilerNode2.
 
+% If compiler mixing is off for a language, each package that uses a
+% compiler within the root unification set "registers" that compiler
+% (and the root set is prohibited from having >= 2 compilers for that
+% language).
+unification_set_compiler("root", CompilerName, CompilerVersion, Language) :-
```

Comment:
If I'm reading correctly, this allows using:
```
CompilerName@CompilerVersion +foo
CompilerName@CompilerVersion ~foo
```
as compilers when the `compiler_mixing` option is on. It would probably be a rare occurrence, but it's worth noting that somewhere if we keep the encoding like it is.

Reply:
Originally I was doing `node()` which seemed to work in a non-reuse context (1st 4 commits of this PR); for preventing mixing during reuse I was not able to identify compilers by `node()` or `hash`. In other words, what is generated now is something I can consistently generate across reused, external, and built packages which prevents mixing in all those cases (but does have an issue if two compilers that differ only by variants exist in the config).